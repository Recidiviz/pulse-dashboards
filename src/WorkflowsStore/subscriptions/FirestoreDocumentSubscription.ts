// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import * as Sentry from "@sentry/react";
import {
  DocumentData,
  DocumentReference,
  DocumentSnapshot,
  onSnapshot,
  Unsubscribe,
} from "firebase/firestore";
import { action, makeObservable, observable, runInAction } from "mobx";

import { FeatureGateError } from "../../errors";
import { castToError } from "../../utils/castToError";
import { FirestoreSubscription } from "./FirestoreSubscription";
import {
  DocumentSubscription,
  TransformFunction,
  UpdateFunction,
  ValidateFunction,
} from "./types";
import {
  defaultTransformFunction,
  defaultUpdateFunction,
  defaultValidateFunction,
} from "./utils";

export abstract class FirestoreDocumentSubscription<
    DataFormat extends DocumentData = DocumentData
  >
  extends FirestoreSubscription<DataFormat | undefined>
  implements DocumentSubscription<DataFormat>
{
  abstract readonly dataSource: DocumentReference;

  constructor(
    // default passes through raw record, assuming it already conforms to the desired format
    protected transformRecord: TransformFunction<DataFormat> = defaultTransformFunction,
    protected validateRecord: ValidateFunction<DataFormat> = defaultValidateFunction,
    protected updateRecord: UpdateFunction<DocumentData> = defaultUpdateFunction
  ) {
    super(undefined, observable.struct);

    // note that dataSource is not observable by default.
    // in the base case there is really no need for it
    makeObservable<this, "updateData">(this, {
      updateData: action,
    });
  }

  /**
   * Stores data on the observable `this.data` property. Mainly intended
   * to be used by the callback for the listener created by `this.subscribe`.
   * Validates the record and does not set `this.data` property if the result
   * fails validation.
   */
  protected updateData(snapshot: DocumentSnapshot): void {
    try {
      const snapshotData = snapshot.data({ serverTimestamps: "estimate" });

      if (!snapshotData) {
        this.data = undefined;
        this.hydrationState = { status: "hydrated" };
        return;
      }

      this.updateRecord(snapshotData);

      const record = this.transformRecord(snapshotData);

      if (record !== undefined) this.validateRecord(record);

      this.data = record;
      this.hydrationState = { status: "hydrated" };
    } catch (e) {
      this.hydrationState = { status: "failed", error: castToError(e) };
      this.data = undefined;
      // don't log routine feature flag checks, but do log everything else
      if (!(e instanceof FeatureGateError)) {
        Sentry.captureException(e);
      }
    }
  }

  protected startSnapshotListener(): Unsubscribe | undefined {
    return onSnapshot(
      this.dataSource,
      (result) => this.updateData(result),
      (error) => {
        runInAction(() => {
          this.hydrationState = { status: "failed", error };
        });
        Sentry.captureException(error);
      }
    );
  }
}
