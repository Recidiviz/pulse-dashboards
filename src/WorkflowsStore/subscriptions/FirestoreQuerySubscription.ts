// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import * as Sentry from "@sentry/react";
import {
  DocumentData,
  onSnapshot,
  Query,
  QuerySnapshot,
  Unsubscribe,
} from "firebase/firestore";
import {
  action,
  computed,
  makeObservable,
  observable,
  onBecomeObserved,
  onBecomeUnobserved,
  reaction,
  runInAction,
} from "mobx";
import { IDisposer } from "mobx-utils";

import {
  QuerySubscription,
  TransformFunction,
  ValidateFunction,
} from "./types";
import { defaultTransformFunction, defaultValidateFunction } from "./utils";

export abstract class FirestoreQuerySubscription<
  DataFormat extends DocumentData
> implements QuerySubscription<DataFormat>
{
  data: DataFormat[] = [];

  abstract get dataSource(): Query | undefined;

  cancelSnapshotListener?: Unsubscribe;

  disposeDynamicDataSource?: IDisposer;

  transformRecord: TransformFunction<DataFormat>;

  validateRecord: ValidateFunction<DataFormat>;

  constructor(
    transformFunction: TransformFunction<DataFormat> = defaultTransformFunction,
    validateFunction: ValidateFunction<DataFormat> = defaultValidateFunction
  ) {
    makeObservable<this, "updateData">(this, {
      data: observable,
      dataSource: computed,
      updateData: action,
      isLoading: observable,
      isHydrated: observable,
      error: observable,
      subscribe: action,
      resetHydration: action,
    });

    this.transformRecord = transformFunction;

    this.validateRecord = validateFunction;

    onBecomeObserved(this, "data", () => this.subscribe());
    onBecomeUnobserved(this, "data", () => this.unsubscribe());
  }

  protected updateData(snapshot: QuerySnapshot | undefined): void {
    const docs: DataFormat[] = [];

    const errors: unknown[] = [];

    snapshot?.forEach((doc) => {
      try {
        const record = this.transformRecord(
          doc.data({ serverTimestamps: "estimate" })
        );

        if (record !== undefined) this.validateRecord(record);

        if (record !== undefined) {
          docs.push(record);
        }
      } catch (e) {
        // Move on to allow for partial success
        errors.push(e);
        Sentry.captureException(e);
      }
    });

    this.data = docs;
    this.error = errors.length ? new AggregateError(errors) : undefined;

    this.isHydrated = true;
    this.isLoading = false;
  }

  get isActive(): boolean {
    return this.cancelSnapshotListener !== undefined;
  }

  subscribe(): void {
    // `subscribe`, on its own, will not re-subscribe if a listener already exists;
    // replacing the subscription with a new one requires further orchestration,
    // as seen in the reaction defined below.
    if (this.isActive || this.isLoading) return;

    this.isLoading = true;

    if (!this.dataSource) {
      this.updateData(undefined);
    } else {
      this.cancelSnapshotListener = onSnapshot(
        this.dataSource,
        (result) => this.updateData(result),
        (error) => {
          runInAction(() => {
            // an error that occurs after data is fetched is considered a partial failure,
            // but this one means that the fetch operation itself failed and therefore so does hydration.
            this.error = error;
            this.isHydrated = false;
            this.isLoading = false;
            this.data = [];
          });
          Sentry.captureException(error);
        }
      );
    }

    // watches dataSource for changes and re-subscribes accordingly
    this.disposeDynamicDataSource = reaction(
      () => this.dataSource,
      () => {
        this.unsubscribe();
        this.resetHydration();
        this.subscribe();
      }
    );
  }

  unsubscribe(): void {
    this.cancelSnapshotListener?.();
    this.cancelSnapshotListener = undefined;
    this.disposeDynamicDataSource?.();
  }

  hydrate(): void {
    this.subscribe();
  }

  isLoading: boolean | undefined = undefined;

  error: Error | undefined = undefined;

  isHydrated = false;

  resetHydration(): void {
    this.isLoading = false;
    this.isHydrated = false;
    this.error = undefined;
  }
}
