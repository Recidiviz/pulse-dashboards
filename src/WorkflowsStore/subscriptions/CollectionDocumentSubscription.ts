// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import {
  doc,
  DocumentData,
  DocumentReference,
  DocumentSnapshot,
} from "firebase/firestore";
import { action, makeObservable, observable, override } from "mobx";

import { CollectionName, collectionNames, db } from "../../firestore";
import { FirestoreDocumentSubscription } from "./FirestoreDocumentSubscription";

/**
 * Subscribes to the specified document in the specified collection.
 * RecordType is essentially just an assertion for downstream convenience;
 * the data source itself is not validated or even strongly typed. Proceed with caution!
 */
export class CollectionDocumentSubscription<
  RecordType extends DocumentData
> extends FirestoreDocumentSubscription<RecordType> {
  readonly dataSource: DocumentReference<RecordType>;

  isLoading?: boolean;

  constructor(collectionName: CollectionName, recordId: string) {
    super();

    makeObservable(this, {
      isLoading: observable,
      subscribe: action,
      updateData: override,
    });

    this.dataSource = doc(
      db,
      collectionNames[collectionName],
      recordId
    ) as DocumentReference<RecordType>;
  }

  subscribe(): void {
    if (this.isLoading === undefined) {
      this.isLoading = true;
    }
    super.subscribe();
  }

  updateData(snapshot: DocumentSnapshot<RecordType>): void {
    super.updateData(snapshot);
    if (this.isLoading) {
      this.isLoading = false;
    }
  }
}
