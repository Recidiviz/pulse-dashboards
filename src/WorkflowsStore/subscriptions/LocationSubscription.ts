// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { collection, Query, query, where } from "firebase/firestore";

import { LocationRecord } from "../../FirestoreStore";
import { FIRESTORE_COLLECTIONS_MAP } from "../../FirestoreStore/constants";
import { RootStore } from "../../RootStore";
import { FirestoreQuerySubscription } from "./FirestoreQuerySubscription";

export class LocationSubscription extends FirestoreQuerySubscription<LocationRecord> {
  private rootStore: RootStore;

  constructor(rootStore: RootStore) {
    super();
    this.rootStore = rootStore;
  }

  get dataSource(): Query {
    const { searchField } = this.rootStore.workflowsStore;

    const stateCode = this.rootStore.currentTenantId;

    const constraints = [
      where("stateCode", "==", stateCode),
      where("idType", "==", searchField),
    ];

    return query(
      collection(
        this.rootStore.firestoreStore.db,
        FIRESTORE_COLLECTIONS_MAP.locations
      ),
      ...constraints
    );
  }
}
