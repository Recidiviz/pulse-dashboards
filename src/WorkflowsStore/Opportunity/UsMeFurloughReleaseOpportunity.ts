// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { differenceInMonths } from "date-fns";
import { computed, makeObservable } from "mobx";

import { WORKFLOWS_METHODOLOGY_URL } from "../../core/utils/constants";
import { OpportunityProfileModuleName } from "../../core/WorkflowsClientProfile/OpportunityProfile";
import { Resident } from "../Resident";
import { OTHER_KEY } from "../utils";
import { OpportunityBase } from "./OpportunityBase";
import { OpportunityRequirement } from "./types";
import {
  UsMeFurloughReleaseReferralRecord,
  usMeFurloughReleaseSchema,
} from "./UsMeFurloughReleaseReferralRecord";
import { CriteriaCopy, CriteriaFormatters, hydrateCriteria } from "./utils";

const DENIAL_REASONS_MAP = {
  "CASE PLAN": "Not compliant with case plan goals",
  PROGRAM: "Has not completed required core programming",
  DISCIPLINE: "Has a Class A or B disciplinary violation pending",
  DECLINE: "Resident declined opportunity to apply for SCCP",
  [OTHER_KEY]: "Other, please specify a reason",
};

const CRITERIA_COPY: CriteriaCopy<UsMeFurloughReleaseReferralRecord> = {
  eligibleCriteria: [
    [
      "usMeServed30DaysAtEligibleFacilityForFurloughOrWorkRelease",
      {
        text: "Served at least 30 days at current facility",
        tooltip:
          "Served at least thirty (30) days of the term of imprisonment in the facility providing the furlough program",
      },
    ],
    [
      "usMeMinimumOrCommunityCustody",
      {
        text: "Currently on $CUSTODY_LEVEL",
        tooltip: "Currently on minimum or community custody",
      },
    ],
    [
      "usMeXMonthsRemainingOnSentence",
      {
        text: "$MONTHS_REMAINING months remaining on sentence",
        tooltip:
          "No more than thirty (30) months remaining on the term of imprisonment or, " +
          "in the case of a split sentence, on the unsuspended portion, after consideration " +
          "of any deductions that the resident has received and retained under Title 17-A, " +
          "Sections 2302(1), 2305, and 23017-2311, if the commissioner, or designee, determines " +
          "that the average statewide case load is no more than ninety (90) adult community " +
          "corrections clients to one probation officer.",
      },
    ],
    [
      "usMeNoClassAOrBViolationFor90Days",
      {
        text: "No Class A or B disciplines pending or occurring in the past 90 days",
        tooltip:
          "Must not have been found guilty of a Class A or B disciplinary violation within ninety " +
          "(90) days of submitting the plan to be transferred to supervised community confinement " +
          "or anytime thereafter prior to the scheduled transfer and must not have a Class A or B " +
          "disciplinary report pending at the time of submitting the plan or scheduled transfer.",
      },
    ],
    [
      "usMeNoDetainersWarrantsOrOther",
      {
        text: "No detainers, warrants, or other pending holds",
        tooltip:
          "Must have no detainers, warrants, or other pending holds preventing participation in a " +
          "community program as set out in Department Policy (AF) 23.1",
      },
    ],
    [
      "usMeServedHalfOfSentence",
      {
        text: "TBD",
        tooltip: "TBD",
      },
    ],
  ],
  ineligibleCriteria: [],
};

const CRITERIA_FORMATTERS: CriteriaFormatters<UsMeFurloughReleaseReferralRecord> =
  {
    eligibleCriteria: {
      usMeMinimumOrCommunityCustody: {
        CUSTODY_LEVEL: ({ custodyLevel }) => custodyLevel.toLowerCase(),
      },
      usMeXMonthsRemainingOnSentence: {
        MONTHS_REMAINING: ({ eligibleDate }) => {
          const monthsRemaining =
            differenceInMonths(eligibleDate, new Date()) + 30;
          return `${monthsRemaining}`;
        },
      },
    },
  } as const;

export class UsMeFurloughReleaseOpportunity extends OpportunityBase<
  Resident,
  UsMeFurloughReleaseReferralRecord
> {
  resident: Resident;

  policyOrMethodologyUrl = WORKFLOWS_METHODOLOGY_URL.US_ME;

  readonly opportunityProfileModules: OpportunityProfileModuleName[] = [
    "Incarceration",
    "CaseNotes",
  ];

  constructor(resident: Resident) {
    super(
      resident,
      "usMeFurloughRelease",
      resident.rootStore,
      usMeFurloughReleaseSchema.parse
    );
    this.resident = resident;

    makeObservable(this, {
      requirementsMet: computed,
    });

    this.denialReasonsMap = DENIAL_REASONS_MAP;
  }

  get requirementsMet(): OpportunityRequirement[] {
    return hydrateCriteria(
      this.record,
      "eligibleCriteria",
      CRITERIA_COPY,
      CRITERIA_FORMATTERS
    );
  }
}
