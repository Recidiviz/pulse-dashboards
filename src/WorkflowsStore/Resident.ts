// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { ResidentRecord } from "../FirestoreStore";
import { RootStore } from "../RootStore";
import tenants from "../tenants";
import { JusticeInvolvedPersonBase } from "./JusticeInvolvedPersonBase";
import {
  IncarcerationOpportunityType,
  Opportunity,
  OpportunityFactory,
  UsMeSCCPOpportunity,
  UsMoRestrictiveHousingStatusHearingOpportunity,
} from "./Opportunity";
import { UsMeFurloughReleaseOpportunity } from "./Opportunity/UsMeFurloughReleaseOpportunity";
import { UsMeWorkReleaseOpportunity } from "./Opportunity/UsMeWorkReleaseOpportunity";
import { UsTnCustodyLevelDowngradeOpportunity } from "./Opportunity/UsTnCustodyLevelDowngradeOpportunity";
import { optionalFieldToDate } from "./utils";

const residentialOpportunityConstructors: Record<
  IncarcerationOpportunityType,
  new (c: Resident) => Opportunity<Resident>
> = {
  usMeFurloughRelease: UsMeFurloughReleaseOpportunity,
  usMeSCCP: UsMeSCCPOpportunity,
  usMeWorkRelease: UsMeWorkReleaseOpportunity,
  usMoRestrictiveHousingStatusHearing:
    UsMoRestrictiveHousingStatusHearingOpportunity,
  usTnCustodyLevelDowngrade: UsTnCustodyLevelDowngradeOpportunity,
};

const createResidentOpportunity: OpportunityFactory<
  IncarcerationOpportunityType,
  Resident
> = (type, person) => {
  return new residentialOpportunityConstructors[type](person);
};

export class Resident extends JusticeInvolvedPersonBase<ResidentRecord> {
  constructor(record: ResidentRecord, rootStore: RootStore) {
    super(record, rootStore, createResidentOpportunity);
  }

  get facilityId(): string | undefined {
    return this.record.facilityId;
  }

  get unitId(): string | undefined {
    return this.record.unitId;
  }

  get custodyLevel(): string | undefined {
    return this.record.custodyLevel;
  }

  get admissionDate(): Date | undefined {
    return optionalFieldToDate(this.record.admissionDate);
  }

  get releaseDate(): Date | undefined {
    return optionalFieldToDate(this.record.releaseDate);
  }

  get searchIdValue(): any {
    const { currentTenantId } = this.rootStore;
    const searchField =
      currentTenantId &&
      tenants[currentTenantId]?.workflowsSystemConfigs?.INCARCERATION
        ?.searchField;

    return searchField ? this.record[searchField] : this.assignedStaffId;
  }

  get portionServedNeeded(): string | undefined {
    return this.record.portionServedNeeded;
  }
}
