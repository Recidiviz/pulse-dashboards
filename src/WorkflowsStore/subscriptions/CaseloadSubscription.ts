// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import {
  collection,
  DocumentData,
  Query,
  query,
  where,
} from "firebase/firestore";

import { CollectionName, collectionNames } from "../../FirestoreStore";
import { WorkflowsStore } from "../WorkflowsStore";
import { FirestoreQuerySubscription } from "./FirestoreQuerySubscription";

export class CaseloadSubscription<
  RecordType extends DocumentData
> extends FirestoreQuerySubscription<RecordType> {
  workflowsStore: WorkflowsStore;

  collectionId: string;

  personType: RecordType["personType"];

  constructor(
    workflowsStore: WorkflowsStore,
    collectionKey: CollectionName,
    personType: RecordType["personType"]
  ) {
    super();
    this.workflowsStore = workflowsStore;
    this.collectionId = collectionNames[collectionKey];
    this.personType = personType;
  }

  get dataSource(): Query | undefined {
    const {
      selectedOfficerIds,
      rootStore: { currentTenantId, firestoreStore },
    } = this.workflowsStore;

    if (!currentTenantId || !selectedOfficerIds.length) {
      return undefined;
    }

    const { collectionId, personType } = this;
    return query(
      collection(firestoreStore.db, collectionId),
      where("stateCode", "==", currentTenantId),
      where("officerId", "in", selectedOfficerIds)
    ).withConverter({
      fromFirestore(snapshot, options) {
        const doc = snapshot.data(options);
        return { ...doc, recordId: snapshot.id, personType };
      },
      // these collections are read-only, so this should never be used, but it is required by Firestore
      toFirestore(record: any) {
        throw new Error(`Writing to ${collectionId} is not allowed`);
      },
    });
  }
}
