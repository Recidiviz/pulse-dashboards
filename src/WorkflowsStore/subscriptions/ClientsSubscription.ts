// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { collection, Query, query, where } from "firebase/firestore";

import { ClientRecord, collectionNames, db } from "../../firestore";
import { WorkflowsStore } from "../WorkflowsStore";
import { FirestoreQuerySubscription } from "./FirestoreQuerySubscription";

export class ClientsSubscription extends FirestoreQuerySubscription<ClientRecord> {
  workflowsStore: WorkflowsStore;

  constructor(workflowsStore: WorkflowsStore) {
    super();
    this.workflowsStore = workflowsStore;
  }

  get dataSource(): Query | undefined {
    const {
      selectedOfficerIds,
      rootStore: { currentTenantId },
    } = this.workflowsStore;

    if (!currentTenantId || !selectedOfficerIds.length) {
      return undefined;
    }

    return query(
      collection(db, collectionNames.clients),
      where("stateCode", "==", currentTenantId),
      where("officerId", "in", selectedOfficerIds)
    ).withConverter({
      fromFirestore(snapshot, options) {
        const doc = snapshot.data(options);
        return { ...doc, recordId: snapshot.id };
      },
      // these collections are read-only, so this should never be used, but it is required by Firestore
      toFirestore(record: any) {
        throw new Error(`Writing to ${collectionNames.clients} is not allowed`);
      },
    });
  }
}
