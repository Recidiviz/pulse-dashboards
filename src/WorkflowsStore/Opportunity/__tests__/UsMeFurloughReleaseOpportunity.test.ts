// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { configure } from "mobx";
import tk from "timekeeper";

import { RootStore } from "../../../RootStore";
import { Resident } from "../../Resident";
import { DocumentSubscription } from "../../subscriptions";
import {
  usMeFurloughReleaseEligibleRecordFixture,
  usMePersonRecord,
} from "../__fixtures__";
import { UsMeFurloughReleaseOpportunity } from "../UsMeFurloughReleaseOpportunity";

let opp: UsMeFurloughReleaseOpportunity;
let resident: Resident;
let root: RootStore;
let referralSub: DocumentSubscription<any>;

jest.mock("../../subscriptions");

function createTestUnit(residentRecord: typeof usMePersonRecord) {
  root = new RootStore();
  jest
    .spyOn(root.workflowsStore, "opportunityTypes", "get")
    .mockReturnValue(["usMeFurloughRelease"]);
  resident = new Resident(residentRecord, root);

  const maybeOpportunity = resident.potentialOpportunities.usMeFurloughRelease;

  if (maybeOpportunity === undefined) {
    throw new Error("Unable to create opportunity instance");
  }

  opp = maybeOpportunity;
}

beforeEach(() => {
  // this lets us spy on observables, e.g. computed getters
  configure({ safeDescriptors: false });
  tk.freeze(new Date(2022, 12, 1));
});

afterEach(() => {
  tk.reset();
  configure({ safeDescriptors: true });
});

describe("fully eligible", () => {
  beforeEach(() => {
    createTestUnit(usMePersonRecord);

    referralSub = opp.referralSubscription;
    referralSub.isLoading = false;
    referralSub.data = usMeFurloughReleaseEligibleRecordFixture;
  });

  test("requirements met", () => {
    expect(opp.requirementsMet).toMatchSnapshot();
  });
});
