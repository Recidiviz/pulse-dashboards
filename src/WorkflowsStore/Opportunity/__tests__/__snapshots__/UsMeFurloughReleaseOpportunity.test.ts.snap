// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fully eligible requirements met 1`] = `
Array [
  Object {
    "text": "Served at least 30 days at current facility",
    "tooltip": "Served at least thirty (30) days of the term of imprisonment in the facility providing the furlough program",
  },
  Object {
    "text": "Currently on community",
    "tooltip": "Currently on minimum or community custody",
  },
  Object {
    "text": "26 months remaining on sentence",
    "tooltip": "No more than thirty (30) months remaining on the term of imprisonment or, in the case of a split sentence, on the unsuspended portion, after consideration of any deductions that the resident has received and retained under Title 17-A, Sections 2302(1), 2305, and 23017-2311, if the commissioner, or designee, determines that the average statewide case load is no more than ninety (90) adult community corrections clients to one probation officer.",
  },
  Object {
    "text": "No Class A or B disciplines pending or occurring in the past 90 days",
    "tooltip": "Must not have been found guilty of a Class A or B disciplinary violation within ninety (90) days of submitting the plan to be transferred to supervised community confinement or anytime thereafter prior to the scheduled transfer and must not have a Class A or B disciplinary report pending at the time of submitting the plan or scheduled transfer.",
  },
  Object {
    "text": "No detainers, warrants, or other pending holds",
    "tooltip": "Must have no detainers, warrants, or other pending holds preventing participation in a community program as set out in Department Policy (AF) 23.1",
  },
]
`;
